contracts/sdk/indexer/indexer/utils.ts

```diff
@@ -124,97 +62,272 @@ export function loadProgram(
+      context,
+      accountKeys,
+      instruction,
+      innerInstructions
+    );
   } else {
-    for (const log of parsedLog.logs) {
-      await indexParsedLog(db, optionalInfo, slot, parserState, log);
+    /// TODO: test with gumball-machine truncate mode
+    /// TODO: write inner instruction parser
+    console.log("[no outer bgum ix found] Ignoring for now");
+    if (innerInstructions.length) {
+      parseBubblegumInnerInstructions(
+        db,
+        slot,
+        parserState,
+        context,
+        accountKeys,
+        innerInstructions[0].instructions,
+      )
     }
   }
 }
 
-export function handleLogsAtomic(
-  db: NFTDatabaseConnection,
-  logs: Logs,
-  context: Context,
-  parsedState: ParserState,
-  startSeq: number | null = null,
-  endSeq: number | null = null
+export function decodeEventInstructionData(
+  idl: Idl,
+  eventName: string,
+  base58String: string,
 ) {
-  if (logs.err) {
-    return;
+  const rawLayouts: [string, Layout<any>][] = idl.events.map((event) => {
+    let eventTypeDef: IdlTypeDef = {
+      name: event.name,
+      type: {
+        kind: "struct",
+        fields: event.fields.map((f) => {
+          return { name: f.name, type: f.type };
+        }),
+      },
+    };
+    return [event.name, IdlCoder.typeDefLayout(eventTypeDef, idl.types)];
+  });
+  const layouts = new Map(rawLayouts);
+  const buffer = bs58.decode(base58String);
+  const layout = layouts.get(eventName);
+  if (!layout) {
+    console.error("Could not find corresponding layout for event:", eventName);
   }
-  const parsedLogs = parseLogs(logs.logs);
-  if (parsedLogs.length == 0) {
-    return;
+  const data = layout.decode(buffer);
+  return { data, name: eventName };
+}
+
+export function destructureBubblegumMintAccounts(
+  accountKeys: PublicKey[],
+  instruction: CompiledInstruction
+) {
+  return {
+    owner: accountKeys[instruction.accounts[4]],
+    delegate: accountKeys[instruction.accounts[5]],
+    merkleSlab: accountKeys[instruction.accounts[6]],
   }
-  db.connection.db.serialize(() => {
-    db.beginTransaction();
-    for (const parsedLog of parsedLogs) {
-      indexParsedLog(
-        db,
-        { txId: logs.signature, startSeq, endSeq },
-        context.slot,
-        parsedState,
-        parsedLog
-      );
+}
+
+
+type ZippedInstruction = {
+  instructionIndex: number,
+  instruction: CompiledInstruction,
+  innerInstructions: CompiledInnerInstruction[],
+}
+
+/// Similar to `order_instructions` in `/nft_ingester/src/utils/instructions.rs`
+function zipInstructions(
+  instructions: CompiledInstruction[],
+  innerInstructions: CompiledInnerInstruction[],
+): ZippedInstruction[] {
+  const zippedIxs: ZippedInstruction[] = [];
+  let innerIxIndex = 0;
+  for (let instructionIndex = 0; instructionIndex < instructions.length; instructionIndex++) {
+    const innerIxs = [];
+    while (innerIxIndex < innerInstructions.length && innerInstructions[innerIxIndex].index <= instructionIndex) {
```

Comment:
Can you explain this clause: `innerInstructions[innerIxIndex].index <= instructionIndex`

Reply:
got it, I think it might be easier to read if you just group innerInstructions into a hashMap: `iiMap = {index: [innerIx]}`

then build the final list:
```
for (const [i, ix] of tx.instructions.entries()) {
    foo.push({
      index: i,
      outerIx: ix,
      innerIxs: iiMap.get(i),
    })
}
```

something along these lines

Reply:
fixed